src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -0,0 +1,146 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    internal class DefaultProjectSnapshotManager : ProjectSnapshotManager
+    {
+        private readonly Workspace _workspace;
+        private readonly Dictionary<ProjectId, ProjectSnapshot> _projects;
+        private readonly List<WeakReference<DefaultProjectSnapshotListener>> _listeners;
+
+        public DefaultProjectSnapshotManager(Workspace workspace)
+        {
+            _workspace = workspace;
+
+            _projects = new Dictionary<ProjectId, ProjectSnapshot>();
+            _listeners = new List<WeakReference<DefaultProjectSnapshotListener>>();
+
+            // Attaching the event handler inside before initialization prevents re-entrancy without
+            // losing any notifications.
+            _workspace.WorkspaceChanged += Workspace_WorkspaceChanged;
+            InitializeSolution(_workspace.CurrentSolution);
+        }
+
+        public override IReadOnlyList<ProjectSnapshot> Projects => _projects.Values.ToArray();
+
+        public override ProjectSnapshot FindProject(string projectPath)
+        {
+            if (projectPath == null)
+            {
+                throw new ArgumentNullException(nameof(projectPath));
+            }
+
+            foreach (var project in _projects.Values)
+            { 
+                if (string.Equals(projectPath, project.UnderlyingProject.FilePath, StringComparison.OrdinalIgnoreCase))
+                {
+                    return project;
+                }
+            }
+
+            return null;
+        }
+
+        public override ProjectSnapshotListener Subscribe()
+        {
+            var subscription = new DefaultProjectSnapshotListener();
+            _listeners.Add(new WeakReference<DefaultProjectSnapshotListener>(subscription));
+
+            return subscription;
+        }
+
+        private void InitializeSolution(Solution solution)
+        {
+            foreach (var kvp in _projects.ToArray())
+            {
+                _projects.Remove(kvp.Key);
+                NofifyListeners(new ProjectChangeEventArgs(kvp.Value, ProjectChangeKind.Removed));
+            }
+
+            if (solution != null)
+
+            foreach (var project in solution.Projects)
+            {
+                var projectState = new DefaultProjectSnapshot(project);
+                _projects[project.Id] = projectState;
+
+                NofifyListeners(new ProjectChangeEventArgs(projectState, ProjectChangeKind.Added));
+            }
+        }
+
+        private void NofifyListeners(ProjectChangeEventArgs e)
+        {
+            for (var i = 0; i < _listeners.Count; i++)
+            {
+                if (_listeners[i].TryGetTarget(out var listener))
+                {
+                    listener.Notify(e);
+                }
+                else
+                {
+                    _listeners.RemoveAt(i--);
```

Comment:
Love me an `i--` 😉

Reply:
RIP

Reply:
I thought about just going backwards, but I thought you'd enjoy it if I memorialized "the taylor"